src/Streamly/Internal/Data/Fold.hs

```diff
@@ -1226,23 +1448,93 @@ distribute_ fs = Fold step initial extract
 partitionByM :: Monad m
     => (a -> m (Either b c)) -> Fold m b x -> Fold m c y -> Fold m a (x, y)
 partitionByM f (Fold stepL beginL doneL) (Fold stepR beginR doneR) =
-
     Fold step begin done
 
     where
 
-    begin = Tuple' <$> beginL <*> beginR
-    step (Tuple' xL xR) a = do
+    begin = do
+        sL <- beginL
+        sR <- beginR
+        return $ RunBoth sL sR
+
+    step (RunBoth sL sR) a = do
+        r <- f a
+        case r of
+            Left b -> do
+                res <- stepL sL b
+                return
+                  $ Partial
+                  $ case res of
+                        Partial sres -> RunBoth sres sR
+                        Partial1 sres -> RunBoth sres sR
+                        Done bres -> RunRight bres sR
+                        Done1 bres -> RunRight bres sR
```

Comment:
It is complicated to correctly determine leftover, we will have to return leftover only when none of the partitions consumed the element. We will have to use different states depending on whether there was a leftover or not.

Alternatively, we can say that we never return a leftover in this combinator.

Reply:
`distribute`, `partition` type of combinators are perhaps best implemented without leftovers. Maybe we should error out if a fold with leftover is used in these, just like we did with transform? Especially, I am not sure if Partial1 can be handled correctly here.

Reply:
Yeah, implementation details bothered me a lot as well. Initially, `Partial1`  ran `step` again. Basically an infinite loop. But then I changed it to this after some thought.